device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ProxiedSSLSocket.java

```diff
@@ -0,0 +1,187 @@
+            proxyConnectMessage = String.format("CONNECT %s %s\r\nHost: %s\r\nProxy-Authorization: Basic %s\r\n\r\n", hostWithPort, HTTP_VERSION_1_1, hostWithPort, base64EncodedCredentials);
+        }
+        else
+        {
+            proxyConnectMessage = String.format("CONNECT %s %s\r\nHost: %s\r\n\r\n\r\n", hostWithPort, HTTP_VERSION_1_1, hostWithPort);
+        }
+
+
+        byte[] proxyConnectBytes = proxyConnectMessage.getBytes(byteEncoding);
+
+        out.write(proxyConnectBytes);
+        out.flush();
+
+        //Cannot do any buffering while reading, only read what is relevant to the connect response
+        HttpHeaderLineInputStream in = new HttpHeaderLineInputStream(tunnel.getInputStream(), byteEncoding);
+
+        String firstLine = in.readUntilHttpNewline();
+        while (!in.readUntilHttpNewline().equals(""));
+
+        //Expects the same http version in the response as the request
+        if (!firstLine.startsWith(HTTP_VERSION_1_1))
+        {
+            tunnel.close();
+            throw new IOException(String.format("Unable to tunnel through %s:%d.  Expected first response line to start with %s, but proxy returns \"%s\"", host, port, HTTP_VERSION_1_1, firstLine));
+        }
+
+        String[] replyStrParts = firstLine.split(" ");
+        if (replyStrParts.length < 2)
+        {
+            tunnel.close();
+            throw new IOException(String.format("Unable to tunnel through %s:%d. Expected proxy response to CONNECT to contain a space between http version and status code, but was %s", host, port, firstLine));
+        }
+
+        int connectResponseStatusCode;
+        try
+        {
+            connectResponseStatusCode = Integer.parseInt(replyStrParts[1]);
+        }
+        catch (NumberFormatException e)
+        {
+            tunnel.close();
+            throw new IOException(String.format("Unable to tunnel through %s:%d. Expected proxy response to CONNECT to contain a status code but status code could not be parsed. Response was %s", host, port, firstLine));
+        }
+
+        if (connectResponseStatusCode <= 199 || connectResponseStatusCode >= 300)
+        {
+            tunnel.close();
+            throw new IOException(String.format("Unable to tunnel through %s:%d. Expected proxy response to CONNECT to return status code 2XX but status code was %d", host, port, connectResponseStatusCode));
+        }
+
+        log.trace("HTTP proxy responded to connect request with status {}, so the proxy connect was successful", connectResponseStatusCode);
+    }
+
+    private interface ExcludedListMethods
+    {
+        void connect(SocketAddress socketAddress, int timeout) throws IOException;
+        void connect(SocketAddress socketAddress) throws IOException;
+        void close() throws IOException;
+    }
+
+    @AllArgsConstructor
+    class HttpHeaderLineInputStream
```

Comment:
It's not following HTTP protocol.
```
    class HttpHeaderLineInputStream
    {
        private final InputStream inputStream;
        private final Charset charset;
        private final ByteArrayOutputStream httpLineOutputStream;
        private boolean done;
        private boolean firstLine;
        HttpHeaderLineInputStream(InputStream inputStream, Charset charset) {
            this.inputStream = inputStream;
            this.charset = charset;
            this.httpLineOutputStream = new ByteArrayOutputStream();
            this.firstLine = true;
        }

        String readUntilHttpNewline() throws IOException
        {
            if (done) {
                return "";
            }
            
            int crlf = 0;
            while (crlf < 2)
            {
                int i = inputStream.read();
                if (i == -1)
                {
                    inputStream.close();
                    throw new IOException("Unexpected EOF from proxy");
                }
                httpLineOutputStream.write((byte) i);
                if ((i == '\r' && crlf == 0) || (i == '\n' && crlf == 1))
                {
                    crlf++;
                }
            }
            String httpHeaderLine = httpLineOutputStream.toString(charset.name());
            httpLineOutputStream.reset();
            if (httpHeaderLine.length() == 0) {
                if (firstLine) {
                    return readUntilHttpNewline();
                } else {
                    done = true;
                }
            } else {
                firstLine = false;
            }
            return httpHeaderLine;
        }
    }
```

Reply:
I'll have this read until /r/n/r/n instead